File: pump/omnipod-eros/src/main/java/app/aaps/pump/omnipod/eros/util/AapsOmnipodUtil.java
package app.aaps.pump.omnipod.eros.util;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializer;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.format.ISODateTimeFormat;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import app.aaps.core.interfaces.resources.ResourceHelper;
import app.aaps.pump.omnipod.eros.driver.definition.AlertSet;
import app.aaps.pump.omnipod.eros.driver.definition.AlertSlot;
import app.aaps.pump.omnipod.eros.driver.definition.AlertType;
import app.aaps.pump.omnipod.eros.driver.manager.ErosPodStateManager;

/**
 * Created by andy on 4/8/19.
 */
@Singleton
public class AapsOmnipodUtil {

    private final ResourceHelper rh;

    private final Gson gsonInstance = createGson();

    @Inject
    public AapsOmnipodUtil(ResourceHelper rh) {
        this.rh = rh;
    }

    private Gson createGson() {
        GsonBuilder gsonBuilder = new GsonBuilder()
                .registerTypeAdapter(DateTime.class, (JsonSerializer<DateTime>) (dateTime, typeOfSrc, context) ->
                        new JsonPrimitive(ISODateTimeFormat.dateTime().print(dateTime)))
                .registerTypeAdapter(DateTime.class, (JsonDeserializer<DateTime>) (json, typeOfT, context) ->
                        ISODateTimeFormat.dateTime().parseDateTime(json.getAsString()))
                .registerTypeAdapter(DateTimeZone.class, (JsonSerializer<DateTimeZone>) (timeZone, typeOfSrc, context) ->
                        new JsonPrimitive(timeZone.getID()))
                .registerTypeAdapter(DateTimeZone.class, (JsonDeserializer<DateTimeZone>) (json, typeOfT, context) ->
                        DateTimeZone.forID(json.getAsString()));

        return gsonBuilder.create();
    }

    public Gson getGsonInstance() {
        return this.gsonInstance;
    }

    public List<String> getTranslatedActiveAlerts(ErosPodStateManager podStateManager) {
        List<String> translatedAlerts = new ArrayList<>();
        AlertSet activeAlerts = podStateManager.getActiveAlerts();

        for (AlertSlot alertSlot : activeAlerts.getAlertSlots()) {
            translatedAlerts.add(translateAlertType(podStateManager.getConfiguredAlertType(alertSlot)));
        }
        return translatedAlerts;
    }

    private String translateAlertType(AlertType alertType) {
        if (alertType == null) {
            return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_unknown_alert);
        }
        switch (alertType) {
            case FINISH_PAIRING_REMINDER:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_finish_pairing_reminder);
            case FINISH_SETUP_REMINDER:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_finish_setup_reminder_reminder);
            case EXPIRATION_ALERT:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_expiration);
            case EXPIRATION_ADVISORY_ALERT:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_expiration_advisory);
            case SHUTDOWN_IMMINENT_ALARM:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_shutdown_imminent);
            case LOW_RESERVOIR_ALERT:
                return rh.gs(app.aaps.pump.omnipod.common.R.string.omnipod_common_alert_low_reservoir);
            default:
                return alertType.name();
        }
    }
}

File: pump/omnipod/eros/src/main/java/app/aaps/pump/omnipod/eros/util/AapsOmnipodUtil.kt
package app.aaps.pump.omnipod.eros.util

import app.aaps.core.interfaces.resources.ResourceHelper
import app.aaps.pump.omnipod.common.R
import app.aaps.pump.omnipod.eros.driver.definition.AlertType
import app.aaps.pump.omnipod.eros.driver.manager.ErosPodStateManager
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializer
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializer
import org.joda.time.DateTime
import org.joda.time.DateTimeZone
import org.joda.time.format.ISODateTimeFormat
import dev.zacsweers.metro.Inject
import dev.zacsweers.metro.AppScope
import dev.zacsweers.metro.SingleIn

/**
 * Created by andy on 4/8/19. Was Java.
 *
 * Gson stays: this is JVM only code and the pod state it serialises is an existing stored format, so
 * swapping the serializer would change what is written to disk.
 *
 * javax `@Singleton`, not Metro's `@SingleIn`: Dagger owns this and hands it to Metro via `PumpLeaves`.
 */
@SingleIn(AppScope::class)
class AapsOmnipodUtil @Inject constructor(
    private val rh: ResourceHelper
) {

    // Public val, not a private val plus a getter: Kotlin callers use `aapsOmnipodUtil.gsonInstance`
    // and Java callers use `getGsonInstance()`. A public val compiles to exactly that getter, so both work.
    val gsonInstance: Gson = createGson()

    private fun createGson(): Gson =
        GsonBuilder()
            .registerTypeAdapter(
                DateTime::class.java,
                JsonSerializer<DateTime> { dateTime, _, _ -> JsonPrimitive(ISODateTimeFormat.dateTime().print(dateTime)) }
            )
            .registerTypeAdapter(
                DateTime::class.java,
                JsonDeserializer { json, _, _ -> ISODateTimeFormat.dateTime().parseDateTime(json.asString) }
            )
            .registerTypeAdapter(
                DateTimeZone::class.java,
                JsonSerializer<DateTimeZone> { timeZone, _, _ -> JsonPrimitive(timeZone.id) }
            )
            .registerTypeAdapter(
                DateTimeZone::class.java,
                JsonDeserializer { json, _, _ -> DateTimeZone.forID(json.asString) }
            )
            .create()

    fun getTranslatedActiveAlerts(podStateManager: ErosPodStateManager): List<String> =
        podStateManager.activeAlerts.alertSlots.map { alertSlot ->
            translateAlertType(podStateManager.getConfiguredAlertType(alertSlot))
        }

    private fun translateAlertType(alertType: AlertType?): String = when (alertType) {
        null                                  -> rh.gs(R.string.omnipod_common_alert_unknown_alert)
        AlertType.FINISH_PAIRING_REMINDER     -> rh.gs(R.string.omnipod_common_alert_finish_pairing_reminder)
        AlertType.FINISH_SETUP_REMINDER       -> rh.gs(R.string.omnipod_common_alert_finish_setup_reminder_reminder)
        AlertType.EXPIRATION_ALERT            -> rh.gs(R.string.omnipod_common_alert_expiration)
        AlertType.EXPIRATION_ADVISORY_ALERT   -> rh.gs(R.string.omnipod_common_alert_expiration_advisory)
        AlertType.SHUTDOWN_IMMINENT_ALARM     -> rh.gs(R.string.omnipod_common_alert_shutdown_imminent)
        AlertType.LOW_RESERVOIR_ALERT         -> rh.gs(R.string.omnipod_common_alert_low_reservoir)
        else                                  -> alertType.name
    }
}
